### DISCUSSION TITLE
How to reset the history of a TS model without to change the model's parameters

### DISCUSSION THREAD

**Discussion by corradomio** [alan-turing-institute/sktime#5545] [General]
There an 'official' mechanism to delete the content of internal _X, _y, _fh, without to change the the model's parameters?

The idea is this:
I use the 2000-2010 data to predict the 2011.
Now I save the trained model inside the DB.
Next I load the model from th DB, I 'update' it with the data 2002-2012 and I use it to predict the 2013.

I don't need to save the historical data because I pass these values before to use it.

In 'theory', it is enough to extend 'model.reset()' with a parameter 'keep_params' with default value 'False' (or other equivalent idea)

**Comment by fkiraly** [comment]
If you explicitly want to make the model "forget" and refit it, you can use `UpdateRefitsEvery` (from `sktime.forecasting.stream`), this will regularly carry out a refit with a fixed lookback window. Is that what you are looking for?

This will not reset `fh` though, in general, as you are using `update` rather than `fit` from the second update on.

The `update` method has sth similar to your suggested `keep_params`, it is the `update_params` parameter.

**Comment by fkiraly** [comment]
Or, are you asking to make the serialized version smaller?

**Comment by corradomio** [comment]
I don't know UpdateRefitsEvery, I will check.
BUT, the idea is NOT to refit the model, the model doesn't change, but only to remove the data that it is not necessary, because replaced in each prediction.
That is, my prediction 'interface' is:

m.predict(fh [,X], y_past [,X_past])

(indeed, now it is implemented using 'update&predict').

The other important reason is, exactly, to reduce the size of the serialized object saved in the DB.

Note: the parameter name I suggested is based on 'update_params' ;-)

Note/2: I implemented a 'clear_yX(model)' based on reflection and some tricks, but it is just a 'trick'.

**Reply by fkiraly** [reply]
hm, care to share your code?
You can open a draft PR for inspection (mark it with "do not merge"), I really would like to understand what you are doing precisely, even if we do not merge it.

**Comment by corradomio** [comment]
```

def _has_xyfh(model):
    return hasattr(model, "_X") and hasattr(model, "_y") and hasattr(model, "_fh")


def clear_yX(model):
    if isinstance(model, list):
        estimators_list = model
        for estimator in estimators_list:
            clear_yX(estimator)

    elif isinstance(model, dict):
        estimators_dict = model
        for key in estimators_dict:
            clear_yX(estimators_dict[key])

    elif _has_xyfh(model):
        model._X = None
        model._y = None

        for attr in [
            "estimators", "_estimators", "estimators_",
            "forecasters", "_forecasters", "forecasters_",
            "estimator", "_estimator", "estimator_",
            "forecaster", "_forecaster", "forecaster_",
        ]:
            if hasattr(model, attr):
                clear_yX(getattr(model, attr))
    return
# end
```
This is just a first implementation, very 'simple'.
I am not sure if it works correctly.
However, for now it is enough ;-)

**Comment by fkiraly** [comment]
Addressed by https://github.com/sktime/sktime/pull/5676, except in case of clones and vectorization.

Review would be appreciated, @corradomio.

**Reply by fkiraly** [reply]
FYI re clone discussion here: https://github.com/sktime/skbase/pull/257

would appreciate your comments, @corradomio

**Comment by fkiraly** [comment]
@corradomio, your code has helped in decoupling the data stores in #5676.

However, I noticed that `self._yvec` can also contain data, through `VectorizedDF`, and I hit a conceptual wall there for now.
See https://github.com/sktime/sktime/pull/5681

Have you also thought about how to handle `self._yvec`?

Have you though about how to do the same in case of vecorization?

**Comment by fkiraly** [comment]
@corradomio, this should be possible for forecasters now, in 0.25.1, via the `remember_data` config. Release notes give you credit for the feature (original idea, code snippet). The feature does not "reset" the history, but does not store `_X`, `_y` in the first place.

As stated above, we did not figure out how to do it for vectorization, so if you have any ideas or code that would be appreciated.
